### PR TITLE
[CARBONDATA-3305] Support show metacache command to list the cache sizes for all tables

### PR DESCRIPTION
**Pull request by QiangCai**
```
>>> SHOW METACACHE
+--------+--------+----------+------------+---------------+
|Database|Table   |Index size|Datamap size|Dictionary size|
+--------+--------+----------+------------+---------------+
|ALL     |ALL     |842 Bytes |982 Bytes   |80.34 KB       |
|default |ALL     |842 Bytes |982 Bytes   |80.34 KB       |
|default |t1      |225 Bytes |982 Bytes   |0              |
|default |t1_dpagg|259 Bytes |0           |0              |
|default |t2      |358 Bytes |0           |80.34 KB       |
+--------+--------+----------+------------+---------------+


>>> SHOW METACACHE FOR TABLE t1
+----------+---------+----------------------+
|Field     |Size     |Comment               |
+----------+---------+----------------------+
|Index     |225 Bytes|1/1 index files cached|
|Dictionary|0        |                      |
|dpagg     |259 Bytes|preaggregate          |
|dblom     |982 Bytes|bloomfilter           |
+----------+---------+----------------------+


>>> SHOW METACACHE FOR TABLE t2
+----------+---------+----------------------+
|Field     |Size     |Comment               |
+----------+---------+----------------------+
|Index     |358 Bytes|2/2 index files cached|
|Dictionary|80.34 KB |                      |
+----------+---------+----------------------+
```

 - [x] Any interfaces changed?
 no
 - [x] Any backward compatibility impacted?
 no
 - [x] Document update required?
yes, added a new command
 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
          add test case
        - How it is tested? Please attach test report.
           yes
        - Is it a performance related change? Please attach the performance test report.
         no
        - Any additional information to help reviewers in testing this change.
         added
       
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
small changes

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2328/

**Comment by qiuchenjian** [comment]
I think LRU is just an index scheduling algorithm，in the future we may support other algorithm 
the command of "show lru" need be considered
How about "show index" or "show index as lru"

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2551/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10587/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/sql/commands/TestCarbonDataShowLRUCommand.scala [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    assertResult(1)(result2.length)
+
+    checkAnswer(sql("show lru on table lru_db.lru_3"),
+      Row("lru_db", "lru_3", 0L, 0L))
+
+    val result4 = sql("show lru on table default.lru_4").collect()
+    assertResult(1)(result4.length)
+
+    val result5 = sql("show lru on table default.lru_5").collect()
+    assertResult(1)(result5.length)
```
Comment (qiuchenjian) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/sql/commands/TestCarbonDataShowLRUCommand.scala]:
```suggestion
    assertResult(1)(result5.length)
    
    sql("use lru_db")
    val result6 = sql("show lru on table lru_3").collect()
    assertResult(1)(result6.length)
```

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/lru/CarbonDataShowLRUCommand.scala [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            dbDictSize = dbDictSize + cache.getMemorySize
+            val memorySize = tableMapDictSize.get(dictId.get._2)
+            if (memorySize.isEmpty) {
+              tableMapDictSize.put(dictId.get._2, cache.getMemorySize)
+            } else {
+              tableMapDictSize.put(dictId.get._2, memorySize.get + cache.getMemorySize)
+            }
+          }
+        }
+      }
```
Comment (qiuchenjian) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/lru/CarbonDataShowLRUCommand.scala]: Will this pr will support the BloomDataMapCache?

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/lru/CarbonDataShowLRUCommand.scala [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+          if (tablePath.isDefined) {
+            dbIndexSize = dbIndexSize + cache.getMemorySize
+            val memorySize = tableMapIndexSize.get(tablePath.get._2)
+            if (memorySize.isEmpty) {
+              tableMapIndexSize.put(tablePath.get._2, cache.getMemorySize)
+            } else {
+              tableMapIndexSize.put(tablePath.get._2, memorySize.get + cache.getMemorySize)
+            }
+          }
+        } else if (cache.isInstanceOf[AbstractColumnDictionaryInfo]) {
```
Comment (qiuchenjian) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/lru/CarbonDataShowLRUCommand.scala]: Will this pr will support the BloomDataMapCache?

**Comment by xuchuanyin** [comment]
emm, to add a unique grammar only for LRU cache is not proper for further evolution I think. What if in the further we extract the storage for the cache and there are more candidates along with LRU?

How about 'SHOW SYSTEM STATUS' or something like this that provide better abstraction?

**Comment by NamanRastogi** [comment]
Can we use the more general word **Cache** in place of **LRU** everywhere?
LRU is our current caching mechanism, we may change that in future to something else. So it should be like the following to generalize to **cache** instead of one specific implementation of caching. 

* `CarbonDataShowLRUCommand` -> `CarbonShowCacheCommand`
* `CacheProvider.getCarbonLRUCache` -> `CacheProvider.getCache`
* `CarbonLRUCache.getLruCacheMap()` -> `CarbonLRUCache.getCacheMap()`
* `show lru on table lru_db.lru_1` -> `show cache on table lru_db.lru_1`

**Comment by QiangCai** [comment]
@NamanRastogi fixed comments

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2582/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10842/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2813/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/sql/commands/TestCarbonDataShowCacheCommand.scala [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    sql("drop database if exists cache_db cascade").collect()
+    sql("drop database if exists cache_empty_db cascade").collect()
+    sql("create database cache_db").collect()
+    sql("create database cache_empty_db").collect()
+    dropTable
+    sql("use cache_db").collect()
+    sql(
+      """
+        | CREATE TABLE cache_db.cache_1
+        | (empno int, empname String, designation String, doj Timestamp, workgroupcategory int,
```
Comment (dhatchayani) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/sql/commands/TestCarbonDataShowCacheCommand.scala]: please add test cases for the child tables like pre-agg table

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2584/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10844/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2815/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2592/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10852/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2823/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2605/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10864/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2836/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2608/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10867/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2839/

**Comment by NamanRastogi** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2610/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10869/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2841/

**Comment by NamanRastogi** [comment]
Please update the description to

```sql
show metacache
show metacache on table lru_db.lru_2
```

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2612/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10871/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2844/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2614/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10874/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2845/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2626/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10886/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2627/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10887/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2858/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+      val table = sparkSession.sessionState.catalog.listTables(dbName)
+        .find(_.table.equalsIgnoreCase(tableIdentifier.get.table))
+      if (table.isEmpty) {
+        throw new NoSuchTableException(dbName, tableIdentifier.get.table)
+      }
+      showTableCache(sparkSession, table.get)
+    }
+  }
+
+  def humanReadableFormat(size: Long): String = {
```
Comment (QiangCai) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala]: How about to use org.apache.commons.io.FileUtils.byteCountToDisplaySize directly

**Comment by xuchuanyin** [comment]
@QiangCai I‘m wondering how this command will work with the distributed IndexServer in the future.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+import org.apache.carbondata.core.metadata.AbsoluteTableIdentifier
+import org.apache.carbondata.core.metadata.schema.table.{CarbonTable, DataMapSchema}
+import org.apache.carbondata.datamap.bloom.BloomCacheKeyValue
+import org.apache.carbondata.processing.merger.CarbonDataMergerUtil
+
+/**
+ * SHOW CACHE
+ */
+case class CarbonShowCacheCommand(tableIdentifier: Option[TableIdentifier])
+  extends DataCommand {
```
Comment (QiangCai) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala]: better to extend MetadataCommand

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala (hunk 5 above)
Comment (xuchuanyin) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala]: Agree with that. Besides there maybe also exists some similar code in carbon-tool module.

**Comment by QiangCai** [comment]
@xuchuanyin 
just do like now
Maybe Index Server should provide some RESTfull API.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+  }
+
+  override def processData(sparkSession: SparkSession): Seq[Row] = {
+    if (tableIdentifier.isEmpty) {
+      showAllTablesCache(sparkSession)
+    } else {
+      val dbName = CarbonEnv.getDatabaseName(tableIdentifier.get.database)(sparkSession)
+      val table = sparkSession.sessionState.catalog.listTables(dbName)
+        .find(_.table.equalsIgnoreCase(tableIdentifier.get.table))
+      if (table.isEmpty) {
```
Comment (QiangCai) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala]: maybe need to check if it is a carbon table or not

**Comment by QiangCai** [comment]
@NamanRastogi please check these three comments

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala (hunk 5 above)
Comment (NamanRastogi) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala]: Done. Using `byteCountToDisplaySize`

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala (hunk 6 above)
Comment (NamanRastogi) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala]: Done.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala (hunk 7 above)
Comment (NamanRastogi) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/cache/CarbonShowCacheCommand.scala]: Done.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/sql/commands/TestCarbonDataShowCacheCommand.scala (hunk 4 above)
Comment (NamanRastogi) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/sql/commands/TestCarbonDataShowCacheCommand.scala]: its present now.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2629/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2631/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2632/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2634/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10894/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2865/

**Comment by xuchuanyin** [comment]
@QiangCai @NamanRastogi Considering there will be a new feature called IndexServer later in carbon, here I'd suggest to use the name 'IndexCache' instead of 'MetaCache'. For example, 'SHOW IndexCache FOR TABLE t1'.

**Comment by QiangCai** [comment]
@xuchuanyin 
It is not for the index cache only. It will include all caches.

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2636/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10895/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2867/

**Comment by xuchuanyin** [comment]
> @xuchuanyin
> It is not for the index cache only. It will include all caches.

But currently you only deal with the index cache.
Actually 'MetaCache' may cause some confusion for understanding the 'meta'. Does it 'table meta' or something else?

**Comment by QiangCai** [comment]
@xuchuanyin It means the metadata index, for example, min/max, dictionary.

**Comment by QiangCai** [comment]
LGTM

**Comment by xuchuanyin** [comment]
LGTM

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/lru/CarbonDataShowLRUCommand.scala (hunk 3 above)
Comment (NamanRastogi) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/lru/CarbonDataShowLRUCommand.scala]: Yes....it does support bloom

**Pull request closed without merging**